eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/jpa/LegacyJpaEventStorageEngine.java

```diff
@@ -0,0 +1,625 @@
+            );
+
+            GapAwareTrackingToken previousToken = cleanedToken((GapAwareTrackingToken) lastToken);
+
+            List<Object[]> entries = transactionManager.fetchInTransaction(() -> legacyJpaOperations.fetchEvents(
+                    previousToken,
+                    batchSize));
+            return legacyJpaOperations.entriesToEvents(
+                    previousToken, entries, tokenOperations.gapTimeoutThreshold(), lowestGlobalSequence, maxGapOffset
+            );
+        }
+
+        private GapAwareTrackingToken cleanedToken(GapAwareTrackingToken lastToken) {
+            if (lastToken != null && lastToken.getGaps().size() > gapCleaningThreshold) {
+                return tokenOperations.withGapsCleaned(lastToken, indexAndTimestampBetweenGaps(lastToken));
+            }
+            return lastToken;
+        }
+
+        private List<Object[]> indexAndTimestampBetweenGaps(GapAwareTrackingToken lastToken) {
+            return transactionManager.fetchInTransaction(() -> legacyJpaOperations.indexAndTimestampBetweenGaps(
+                    lastToken));
+        }
+
+
+        private boolean defaultFinalAggregateBatchPredicate(List<? extends DomainEventData<?>> recentBatch) {
+            return fetchForAggregateUntilEmpty() ? recentBatch.isEmpty() : recentBatch.size() < batchSize;
+        }
+
+        private static class EventStreamSpliterator<T> extends Spliterators.AbstractSpliterator<T> {
+
+            private final Function<T, List<? extends T>> fetchFunction;
+            private final Predicate<List<? extends T>> finalBatchPredicate;
+
+            private Iterator<? extends T> iterator;
+            private T lastItem;
+            private boolean lastBatchFound;
+
+            private EventStreamSpliterator(Function<T, List<? extends T>> fetchFunction,
+                                           Predicate<List<? extends T>> finalBatchPredicate) {
+                super(Long.MAX_VALUE, NONNULL | ORDERED | DISTINCT | CONCURRENT);
+                this.fetchFunction = fetchFunction;
+                this.finalBatchPredicate = finalBatchPredicate;
+            }
+
+            @Override
+            public boolean tryAdvance(Consumer<? super T> action) {
+                Objects.requireNonNull(action);
+                if (iterator == null || !iterator.hasNext()) {
+                    if (lastBatchFound) {
+                        return false;
+                    }
+                    List<? extends T> items = fetchFunction.apply(lastItem);
+                    lastBatchFound = finalBatchPredicate.test(items);
+                    iterator = items.iterator();
+                }
+                if (!iterator.hasNext()) {
+                    return false;
+                }
+
+                action.accept(lastItem = iterator.next());
+                return true;
+            }
+        }
+    }
+
+
+    public record Customization(
```

Comment:
Not so sure on calling this configuration object `Customization`. **But**, you don't have to solve that now. Just want to share my current opinion. There are more of these to come, so we can always adjust later.

Reply:
I'm not sure too, but it's as we called this one while talking about configuration. 
Let's talk about it later as you say. I also have some more ideas to discuss about the config.